### DISCUSSION TITLE
How to Convert Complex TypeScript Interfaces to Zod Schema?

### DISCUSSION THREAD

**Discussion by alan890104** [colinhacks/zod#3885] [Q&A]
Hi everyone,

I'm working on a TypeScript project and need to convert a set of complex interfaces into a Zod schema for runtime validation. Here's a simplified example of the interfaces I'm working with:

```typescript
export type WithdrawMode = 'Single' | 'Balanced'

interface BaseWithdraw {
  pool: Address
  removeLpAmount: bigint
  queryId?: bigint
  slippageTolerance?: Slippage
  minAmountOuts?: IAllocation[] | IAllocation
  extraPayload?: ExtraPayload
}

// Strictly define NextWithdraw based on mode
interface NextWithdrawSingle {
  pool: Address
  mode: 'Single'
  withdrawAsset: Asset // Must be defined for Single mode
  removeLpAmount?: bigint
}

interface NextWithdrawBalanced {
  pool: Address
  mode: 'Balanced'
  withdrawAsset?: never // Must be undefined for Balanced mode
  removeLpAmount?: bigint
}

type NextWithdraw = NextWithdrawSingle | NextWithdrawBalanced

// Single mode base type
interface SingleWithdrawBase extends BaseWithdraw {
  mode: 'Single'
}

// Mutual exclusivity between withdrawAsset and nextWithdraw
type SingleWithdrawWithNext = SingleWithdrawBase & {
  nextWithdraw: NextWithdraw
  withdrawAsset?: never // Enforce withdrawAsset is undefined when nextWithdraw is defined
}

type SingleWithdrawWithAsset = SingleWithdrawBase & {
  withdrawAsset: Asset
  nextWithdraw?: never // Enforce nextWithdraw is undefined when withdrawAsset is defined
}

type SingleWithdrawParams = SingleWithdrawWithNext | SingleWithdrawWithAsset

// Balanced mode type
interface BalancedWithdrawParams extends BaseWithdraw {
  mode: 'Balanced'
  nextWithdraw?: NextWithdraw // No restrictions for nextWithdraw in Balanced mode
}

// Unified WithdrawParams type
export type WithdrawParams = SingleWithdrawParams | BalancedWithdrawParams
```

My goal is to create Zod schemas for these types, especially for `WithdrawParams`, which involves unions and nested objects. I'm unsure how to handle the more complex parts, like:

- Union types (`SingleWithdrawParams` and `BalancedWithdrawParams`)
- Conditional fields (`withdrawAsset` and `nextWithdraw`)
- Optional fields and nested objects (`NextWithdraw`)

The rules:
- if mode is Single and nextWithdraw is defined, then withdrawAsset must be undefined
- if mode is Single and nextWithdraw is undefined, then withdrawAsset must be defined
- if mode is Balanced, then withdrawAsset must be undefined
- if mode of nextWithdraw is Single, then nextWithdraw.withdrawAsset must be defined
- if mode of nextWithdraw is Balanced, then nextWithdraw.withdrawAsset must be undefined

Could anyone provide guidance or examples of how to translate these interfaces into Zod schemas? Any advice or resources would be greatly appreciated!

Thank you!

**Comment by debel27** [comment]
Haven't tried it, but the [ts-to-zod](https://github.com/fabien0102/ts-to-zod) seems to fit your needs.

I'm sure you're aware, going "types first" means opting out of advanced Zod features, as described in the docs:

> Zod was designed to mirror TypeScript as closely as possible. But there are many so-called "refinement types" you may wish to check for that can't be represented in TypeScript's type system. For instance: checking that a number is an integer or that a string is a valid email address.

To support these use cases, you would have to implement additional logic.
